### PR TITLE
[RFC] Fix const in nested template arg

### LINKED ISSUE #1355: const fails when used as the argument to a template of a template
This works:

```
from libcpp.vector cimport vector

cdef extern from "Test.hpp":
    cdef vector[char](const) f()
```

As does this:

```
from libcpp.vector cimport vector

cdef extern from "Test.hpp":
    cdef vector[f()
```

But this causes a Cython parse error on Cython 0.18:

```
from libcpp.vector cimport vector

cdef extern from "Test.hpp":
    cdef vector[vector[const char](vector[char]])] f()
```

Migrated from http://trac.cython.org/ticket/798

### PR DESCRIPTION
**Pull request by shwina**
Fixes #1355 

As noted in https://github.com/cython/cython/issues/3885, declarations such as the following currently fail:

```
cdef vector[vector[const int]] x
```

When the parser sees `cdef vector[...]`, it calls `looking_at_expr` [here](https://github.com/cython/cython/issues/3885) to figure out if `...` is an expression. Currently, that returns `True`. The approach taken here is to ensure it returns `False` so that `...` is parsed correctly as a _type_. I hope this the right way to go about it.

### CODE REVIEW

File: Cython/Compiler/Parsing.py [hunk 1]
```diff
@@ -2700,6 +2700,10 @@ def looking_at_expr(s):
             s.next()
             is_type = s.sy == ']'
             s.put_back(*saved)
+            if not is_type:  # could be template type
+                s.next()
+                is_type = not looking_at_expr(s)
+                s.put_back(*saved)
```
[the remark this answers is not in the document]
Comment (shwina) [reply Cython/Compiler/Parsing.py:2706]: I think you're absolutely right :)

**Comment by shwina** [comment]
Looks like it worked -- any chance of getting this merged? :-)

**Comment by scoder** [comment]
Thanks!

**Comment by shwina** [comment]
Thanks @scoder! Would it be possible for this to be backported to an upcoming 0.29.x release?

**Comment by jpfeuffer** [comment]
Hi,

we are currently experiencing problems with nested templates and the latest 0.29.22 version. Was this PR tested with spaces between the brackets e.g. `libcpp_vector [ libcpp_vector [ double ] ]` ?
Or is this not valid syntax?

I will investigate more on our side, too.
Thanks!

**Comment by jpfeuffer** [comment]
Ah I think we figured it out. We were kind of working around the const problem which we do not need to anymore. Thanks again.

**Comment by uweschmitt** [comment]
> Hi,
> 
> we are currently experiencing problems with nested templates and the latest 0.29.22 version. Was this PR tested with spaces between the brackets e.g. `libcpp_vector [ libcpp_vector [ double ] ]` ?
> Or is this not valid syntax?
> 
> I will investigate more on our side, too.
> Thanks!

Fixed in autowrap 0.22.2.

**Pull request merged**